### PR TITLE
[JIT] ARM64 - Temporary fix for `ldp`/`stp` optimizations

### LINKED ISSUE #85765: JIT: Invalid ldp optimization with locals
```csharp
// Generated by Fuzzlyn v1.5 on 2023-05-03 21:56:56
// Run on Arm64 Linux
// Seed: 12429971184890412752
// Reduced from 299.4 KiB to 0.4 KiB in 00:02:13
// Debug: Outputs False
// Release: Outputs True
public struct S0
{
    public S0(bool f1): this()
    {
    }
}

public struct S1
{
    public byte F0;
    public bool F1;
    public bool F2;
}

public class Program
{
    public static void Main()
    {
        S1 vr2 = M4();
        vr2.F2 |= vr2.F1;
        System.Console.WriteLine(vr2.F2);
    }

    public static S1 M4()
    {
        S1 var1 = default(S1);
        var vr0 = new S0(false);
        return var1;
    }
}
```

The codegen for `Main` looks odd:
```asm
; Assembly listing for method Program:Main()
; Emitting BLENDED_CODE for generic ARM64 CPU - Windows
; optimized code
; fp based frame
; partially interruptible
; No PGO data
; invoked as altjit
; Final local variable assignments
;
;  V00 loc0         [V00    ] (  5,  5   )  struct ( 8) [fp+18H]   do-not-enreg[SB] ld-addr-op
;# V01 OutArgs      [V01    ] (  1,  1   )  struct ( 0) [sp+00H]   do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;  V02 tmp1         [V02,T03] (  1,  1   )   ubyte  ->  [fp+18H]   do-not-enreg[] V00.F0(offs=0x00) P-DEP "field V00.F0 (fldOffset=0x0)"
;  V03 tmp2         [V03,T02] (  2,  2   )    bool  ->  [fp+19H]   do-not-enreg[] V00.F1(offs=0x01) P-DEP "field V00.F1 (fldOffset=0x1)"
;  V04 tmp3         [V04,T00] (  4,  4   )    bool  ->  [fp+1AH]   do-not-enreg[] single-def V00.F2(offs=0x02) P-DEP "field V00.F2 (fldOffset=0x2)"
;  V05 rat0         [V05,T01] (  2,  4   )  struct ( 8) [fp+10H]   do-not-enreg[SF] "Return value temp for an odd struct return size"
;
; Lcl frame size = 16

G_M27646_IG01:  ;; offset=0000H
        A9BE7BFD          stp     fp, lr, [sp, #-0x20]!
        910003FD          mov     fp, sp
                                                ;; size=8 bbWeight=1 PerfScore 1.50
G_M27646_IG02:  ;; offset=0008H
        D29A6400          movz    x0, #0xD320      // code for Program:M4():S1
        F2AB0D00          movk    x0, #0x5868 LSL #16
        F2CFFF20          movk    x0, #0x7FF9 LSL #32
        F9400000          ldr     x0, [x0]
        D63F0000          blr     x0
        B90013A0          str     w0, [fp, #0x10]
        794023A0          ldrh    w0, [fp, #0x10]
        790033A0          strh    w0, [fp, #0x18]
        39404BA0          ldrb    w0, [fp, #0x12]
        39006BA0          strb    w0, [fp, #0x1A]
        294C83A1          ldp     w1, w0, [fp, #0x64]
        2A010000          orr     w0, w0, w1
        39006BA0          strb    w0, [fp, #0x1A]       // [V04 tmp3]
        39406BA0          ldrb    w0, [fp, #0x1A]       // [V04 tmp3]
        D28DC101          movz    x1, #0x6E08      // code for System.Console:WriteLine(bool)
        F2AB10A1          movk    x1, #0x5885 LSL #16
        F2CFFF21          movk    x1, #0x7FF9 LSL #32
        F9400021          ldr     x1, [x1]
        D63F0020          blr     x1
                                                ;; size=76 bbWeight=1 PerfScore 23.50
G_M27646_IG03:  ;; offset=0054H
        A8C27BFD          ldp     fp, lr, [sp], #0x20
        D65F03C0          ret     lr
                                                ;; size=8 bbWeight=1 PerfScore 2.00

; Total bytes of code 92, prolog size 8, PerfScore 36.20, instruction count 23, allocated bytes for code 92 (MethodHash=cb019401) for method Program:Main()
; ============================================================
```
The:
```asm
        294C83A1          ldp     w1, w0, [fp, #0x64]
```
is loading far outside the stack frame.

From the dump:
```scala
Generating: N021 (  2,  2) [000008] -----------                    t8 =    LCL_VAR   int    V04 tmp3         u:2 x0 (last use) REG x0 $201
IN000b:                           ldr     w0, [fp, #0x1A]
Generating: N023 (  3,  2) [000011] -----------                   t11 =    LCL_VAR   int    V03 tmp2         u:2 x1 (last use) REG x1 $200
Removing saved instruction in current IG G_M27646_IG02:
> IN000b:                           ldr     w0, [fp, #0x1A]
IN000b:                           ldp     w1, w0, [fp, #0x64]
```

Maybe the optimization does not handle unaligned loads from the stack frame correctly? (For some arm64 instructions those are not representable)

cc @dotnet/jit-contrib @kunalspathak

### PR DESCRIPTION
**Pull request by TIHan**
Resolves https://github.com/dotnet/runtime/issues/85765

With the latest, the code-gen is quite different from what was reported in the issue, and therefore doesn't reproduce. But the issue still exists and is able to be reproduced by a different sample:

```csharp
using System;
using System.Runtime.CompilerServices;

// Expected: 515
// Actual: 0
public unsafe class Program
{
    public static void Main()
    {
        byte* bytes = stackalloc byte[1024];
        bytes[0x1A] = 1;
        bytes[0x1B] = 2;
        int sum = Foo(bytes);
        Console.WriteLine(sum);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static int Foo(byte* b)
    {
        return Unsafe.ReadUnaligned<int>(ref b[0x1A]) + Unsafe.ReadUnaligned<int>(ref b[0x1B]);
    }
}
```

### CODE REVIEW

**Comment by jakobbotsch** [comment]
This repros up until #86491 was merged. Can you check out e62cb64335549d48439585d436a4903ab06136a3 and repro the problem there? With altjit and TC=0 I get the same codegen as in the issue on that commit.

**Comment by TIHan** [comment]
@jakobbotsch What did you do to determine which PR could repo this? Did you just do a bisect?

**Comment by jakobbotsch** [comment]
> @jakobbotsch What did you do to determine which PR could repo this? Did you just do a bisect?

Yes. I keep Core_Roots compiled for all JIT commits so that I can quickly do that.

**Comment by TIHan** [comment]
> Yes. I keep Core_Roots compiled for all JIT commits so that I can quickly do that.

That's a lot of Core_Roots :)

**Comment by TIHan** [comment]
I checked out https://github.com/dotnet/runtime/commit/e62cb64335549d48439585d436a4903ab06136a3 and did a fresh/clean checked build and the codegen is the same. Maybe I should try to go back further.

**Comment by TIHan** [comment]
Marking it as NoInline I was able to reproduce it, but only on that commit.

Latest code-gen is:
```asm
; Assembly listing for method Program:Main() (FullOpts)
; Emitting BLENDED_CODE for generic ARM64 - Windows
; FullOpts code
; optimized code
; fp based frame
; partially interruptible
; No PGO data
; invoked as altjit
; Final local variable assignments
;
;  V00 loc0         [V00    ] (  5,  5   )  struct ( 8) [fp+0x18]  do-not-enreg[SB] ld-addr-op <S1>
;# V01 OutArgs      [V01    ] (  1,  1   )  struct ( 0) [sp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;  V02 tmp1         [V02,T03] (  1,  1   )   ubyte  ->  [fp+0x18]  do-not-enreg[] "field V00.F0 (fldOffset=0x0)" P-DEP
;  V03 tmp2         [V03,T02] (  2,  2   )    bool  ->  [fp+0x19]  do-not-enreg[] "field V00.F1 (fldOffset=0x1)" P-DEP
;  V04 tmp3         [V04,T00] (  4,  4   )    bool  ->  [fp+0x1A]  do-not-enreg[] single-def "field V00.F2 (fldOffset=0x2)" P-DEP
;  V05 rat0         [V05,T01] (  2,  4   )  struct ( 8) [fp+0x10]  do-not-enreg[SF] "Return value temp for an odd struct return size" <S1>
;
; Lcl frame size = 16

G_M27646_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x20]!
            mov     fp, sp
						;; size=8 bbWeight=1 PerfScore 1.50
G_M27646_IG02:  ;; offset=0x0008
            movz    x0, #0xC408      // code for Program:M4():S1
            movk    x0, #0x5CDC LSL #16
            movk    x0, #0x7FFD LSL #32
            ldr     x0, [x0]
            blr     x0
            str     w0, [fp, #0x10]	// [V05 rat0]
            ldrh    w0, [fp, #0x10]
            strh    w0, [fp, #0x18]
            ldrb    w0, [fp, #0x12]
            strb    w0, [fp, #0x1A]
            ldrb    w0, [fp, #0x1A]	// [V04 tmp3]
            ldrb    w1, [fp, #0x19]	// [V03 tmp2]
            orr     w0, w0, w1
            strb    w0, [fp, #0x1A]	// [V04 tmp3]
            ldrb    w0, [fp, #0x1A]	// [V04 tmp3]
            movz    x1, #0x4CD8      // code for System.Console:WriteLine(bool)
            movk    x1, #0x5CFF LSL #16
            movk    x1, #0x7FFD LSL #32
            ldr     x1, [x1]
            blr     x1
						;; size=80 bbWeight=1 PerfScore 25.50
G_M27646_IG03:  ;; offset=0x0058
            ldp     fp, lr, [sp], #0x20
            ret     lr
						;; size=8 bbWeight=1 PerfScore 2.00

; Total bytes of code 96, prolog size 8, PerfScore 38.60, instruction count 24, allocated bytes for code 96 (MethodHash=cb019401) for method Program:Main() (FullOpts)
; ============================================================
```

**Comment by jakobbotsch** [comment]
> Marking it as NoInline I was able to reproduce it, but only on that commit.

This is in the backend, so we should fix it even if it no longer repros with this specific example. #86491 is a change in morph so it did not fix the backend bug.

**Comment by TIHan** [comment]
> This is in the backend, so we should fix it even if it no longer repros with this specific example. https://github.com/dotnet/runtime/pull/86491 is a change in morph so it did not fix the backend bug.

Looks like it is, but all I can do is try to fix it based on this commit and hope that it is correct considering there are no other examples that reproduce it in the latest.

**Comment by jakobbotsch** [comment]
> Looks like it is, but all I can do is try to fix it based on this commit and hope that it is correct considering there are no other examples that reproduce it in the latest.

Fixing the backend bug on a commit that is a few months older is just fine. You should be able to test your fix on that commit. I did the same in #90246.

**Comment by TIHan** [comment]
It doesn't leave me feeling confident knowing that this cannot be reproduced in latest.

**Comment by jakobbotsch** [comment]
How would you know that the problem cannot be reproduced in `main`? Once you understand the problem you might even be able to construct an example yourself.

**Comment by jakobbotsch** [comment]
Here is an example that reproduces the problem on main:
```csharp
using System;
using System.Runtime.CompilerServices;

public unsafe class Program
{
    public static void Main()
    {
        byte* bytes = stackalloc byte[1024];
        bytes[0x1A] = 1;
        bytes[0x1B] = 2;
        int sum = Foo(bytes);
        Console.WriteLine(sum);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static int Foo(byte* b)
    {
        return Unsafe.ReadUnaligned<int>(ref b[0x1A]) + Unsafe.ReadUnaligned<int>(ref b[0x1B]);
    }
}
```

Expected: 515
Actual: 0

**Comment by TIHan** [comment]
> Once you understand the problem you might even be able to construct an example yourself.

That's the tricky part for this problem. I have no idea what I'm looking at as it's new to me.

**Comment by TIHan** [comment]
That new example, interestingly, it outputs 515 in https://github.com/dotnet/runtime/commit/e62cb64335549d48439585d436a4903ab06136a3 .

**Comment by jakobbotsch** [comment]
> That new example, interestingly, it outputs 515 in [e62cb64](https://github.com/dotnet/runtime/commit/e62cb64335549d48439585d436a4903ab06136a3) .

Odd, I get the same codegen on main and e62cb64: a `ldp     w1, w0, [x0, #0x68]` that incorrectly loads at 0x1A * 4.

**Comment by TIHan** [comment]
@jakobbotsch I made a quick fix, but I put in a "TODO". 

The problem is a little complicated, but the issue is that 'imm' and/or 'prevImm' are assumed to be "scaled" when attempting to do a `ldr`/`str` pair optimization to `ldp`/`stp` respectively. The bug is that 'imm' and/or 'prevImm' are not "scaled" so it's using the wrong values. We should fix this, but we can fix it later.

**Comment by TIHan** [comment]
@dotnet/jit-contrib @BruceForstall @jakobbotsch this is ready, pending CI.

File: src/tests/JIT/Regression/JitBlue/Runtime_85765/Runtime_85765.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    // ------
+
+    [Fact]
+    public static void Test2()
+    {
+        byte* bytes = stackalloc byte[1024];
+        bytes[0x1A] = 1;
+        bytes[0x1B] = 2;
+        int sum = Foo(bytes);
+        Assert.Equal(515, sum);
```
Comment (markples) [inline src/tests/JIT/Regression/JitBlue/Runtime_85765/Runtime_85765.cs:55]: I believe that the consensus was not to use Equal due to (1) it relies on reflection and is likely to fail in nativeaot scenarios, and (2) it ends up inlined and makes a mess of the assembly to debug.

Assert.True/False would be ok for (1) but probably not (2).

Comment (TIHan) [reply src/tests/JIT/Regression/JitBlue/Runtime_85765/Runtime_85765.cs:55]: Ok, that's interesting. Can change it to Assert.True/False.

**Comment by BruceForstall** [comment]
[No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=375198&view=ms.vss-build-web.run-extensions-tab)

**Review by BruceForstall — APPROVED** [review]
LGTM

We should back-port this to .NET 8.

In the future, I'd like to see the "peephole optimization" code deal in "actual" values, not "encoded" values.

**Comment by jakobbotsch** [comment]
@TIHan The test added here doesn't build -- seems like CI was red when this PR was merged.

**Comment by elinor-fung** [comment]
@TIHan the test added doesn't build: `src/tests/JIT/Regression/JitBlue/Runtime_85765/Runtime_85765.cs(59,27): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context [/__w/1/s/src/tests/JIT/Regression/JitBlue/Runtime_85765/Runtime_85765.csproj]`

**Comment by TIHan** [comment]
Ok, will make a quick PR to fix this.

**Comment by TIHan** [comment]
https://github.com/dotnet/runtime/pull/90698

**Comment by jkotas** [comment]
I am going to revert this. This can be introducing number of other problems since the CI was all read when this was merged.

**Comment by TIHan** [comment]
This is the same PR as this but with the test fix: https://github.com/dotnet/runtime/pull/90700

My fault for not checking.

**Pull request merged**